<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>. - RNA Seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="style.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">.</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../note.html" rel="" target="">
 <span class="menu-text">Note</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../slides.html" rel="" target="">
 <span class="menu-text">Slides</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#序-生物背景" id="toc-序-生物背景" class="nav-link active" data-scroll-target="#序-生物背景">序-生物背景</a></li>
  <li><a href="#实验篇" id="toc-实验篇" class="nav-link" data-scroll-target="#实验篇">实验篇</a></li>
  <li><a href="#rnaseq上游分析篇" id="toc-rnaseq上游分析篇" class="nav-link" data-scroll-target="#rnaseq上游分析篇">RNASeq上游分析篇</a></li>
  <li><a href="#rnaseq下游分析篇" id="toc-rnaseq下游分析篇" class="nav-link" data-scroll-target="#rnaseq下游分析篇">RNASeq下游分析篇</a></li>
  <li><a href="#基因注释与富集分析" id="toc-基因注释与富集分析" class="nav-link" data-scroll-target="#基因注释与富集分析">基因注释与富集分析</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">RNA Seq</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="序-生物背景" class="level2">
<h2 class="anchored" data-anchor-id="序-生物背景">序-生物背景</h2>
<p>组学分析包括基因组学(全外显子组学，全基因组学)，转录组学（有参，无参），表观组学（chip-seq，甲基化等）。</p>
<p>RNA-seq的重要在于，很多情况下疾病的原因是蛋白的原因，而蛋白是由mRNA翻译而来，通过RNA我们可以了解到究竟是哪些蛋白发生发生了变化，或者转录出现了问题。对于组学分析而言，你需要知道哪些物种是已经被测序的，对于已测序的动物，你可以在Ensembl这个网站找到，植物在Ensembl Plant查询。</p>
<p>基因组包括细胞核基因组以及细胞质基因组（主要是线粒体），人基因组大概3.1Gbp，其中coding- area 1.1%，RNA genes以及 regulatory area 3%。线粒体16.6Kb。</p>
<p>基因在染色上分布不均匀。人类大概有20000蛋白基因，7000RNA基因，6号染色体基因数目最多。</p>
<p>rRNA 18 / 28 5.8 5 - 40/ 60。可以从NCBI下载到rRNA序列信息。</p>
<p>tRNA mapping时的参考序列需要加CCA, 组氨酸的参考序列+ G。有的tRNA含有内含子。可以从GtRNAdb下载tRNA序列</p>
<p>mRNA 序列信息可以从Uniprot下载。</p>
<p>microRNA 可以从miRbase中下载。</p>
<p>数量：rRNA(80-90%) tRNA(10%)，mRNA(1-5%)，直接测序大部分都是tRNA，提取mRNA(通过其poly A)，或者去除tRNA再测序。</p>
</section>
<section id="实验篇" class="level2">
<h2 class="anchored" data-anchor-id="实验篇">实验篇</h2>
<section id="rna-测序文库" class="level4">
<h4 class="anchored" data-anchor-id="rna-测序文库">RNA 测序文库</h4>
<p>mRNA 建库</p>
<ol type="1">
<li>mRNA or total RNA</li>
<li>remove contaminant DNA (remove rRNA or select mRNA?)</li>
<li>fragment RNA</li>
<li>reverse transcrible into cDNA（strand specefic?)</li>
<li>ligate sequence adaptors</li>
</ol>
<p>建库时的几个问题。</p>
<ul>
<li><p>是去除tRNA还是特异性保留mRNA？</p>
<p>poly A+ RNA-seq方法通过与mRNA的poly A尾巴特异性结合选择mRNA(无法区分正负链，基因overlap ，正负链都可能是基因)，但是组蛋白RNA没有poly A尾巴。</p>
<p>rRNA - RNA-seq方法通过酶降解tRNA。</p>
<p>两种建库方式的原始数据还是有差异的。</p></li>
<li><p>如何保留链特异性？</p>
<p>通过dUTP method，加入dUTP合成，切割该链条。</p></li>
</ul>
</section>
<section id="测序技术" class="level4">
<h4 class="anchored" data-anchor-id="测序技术">测序技术</h4>
<p>illumina测序原理可以参考这里。</p>
<p>reads1/reads2（/ - 互补） + adapater = fragment<br>
reads1/reads)（/ - 互补) = insert distance</p>
<ul>
<li><p>illumina测不长的原因？</p>
<p>每测一轮都有可能有错误（同一簇不同步），后面越来越差，杂色参杂。</p></li>
<li><p>片段为什么要均一？</p>
<p>短片段更容易在flow cell 结合，导致测出来的都是短片段</p></li>
<li><p>为什么需要保持碱基平衡？</p>
<p>150次flow cell快照，每簇都同一色不好解方程。</p></li>
<li><p>adapter序列出现在结果中？ insert短，没有150bp（只会出现在3端），fasta从左到右边，只有右边有接头；测序是从5端到3端进行测序。</p></li>
</ul>
</section>
<section id="实验数据质控" class="level4">
<h4 class="anchored" data-anchor-id="实验数据质控">实验数据质控</h4>
<p>total RNA提取的质控标准：RIN(RNA Integrity Number)，根据5S, 18, 28S的峰值进行评估，范围0-10。（6-6.5，7）</p>
</section>
</section>
<section id="rnaseq上游分析篇" class="level2">
<h2 class="anchored" data-anchor-id="rnaseq上游分析篇">RNASeq上游分析篇</h2>
<section id="quality-control" class="level4">
<h4 class="anchored" data-anchor-id="quality-control">Quality Control</h4>
<p>质控主要需要考虑一下几个方面：</p>
<ul>
<li>去除adapter</li>
<li>去除低质量reads</li>
<li>去除reads部分低质量区域</li>
<li>为下一步分析做准备；例如研究可变剪切，需要把reads修剪成等长。</li>
</ul>
<p>关于fasta文件：</p>
<p>第一行，<span class="citation" data-cites="reads名字">@reads名字</span>（测序仪编号:lane:tail:x:y)<br>
第二行，序列<br>
第三行，+（reads名字/没有）<br>
第四行，phred value + 33 对应的ASSCI编码<br>
<span class="math inline">\(Phred("F") = 70 - 33 = 37\)</span>(Sanger 标准)<br>
<span class="math inline">\(Phred = -10 * log_{10}(error\ probability)\)</span></p>
<p>Phred40 = 0.0001 error rate<br>
Phred30 = 0.001 error rate<br>
Phred20 = 0.01 error rate<br>
Phred10 = 0.1 error rate</p>
<p>质检报告html图表:</p>
<p>per base quality：总reads每个位置碱基的犯错概率（下限q30）。<br>
per tile sequence quality：如果有花的，那么整个tile可能都有问题。<br>
GCcontent：不同物种GC含量不一样，样品可能被污染。<br>
per base N content: 杂色，未测出碱基是什么。<br>
sequence duplicate level：重复reads展示。<br>
adapter content：序列的接头检测</p>
<p>质控参数：</p>
<p>quality -10</p>
<ol type="1">
<li>Phred - 10</li>
<li>数据从右向左累加</li>
<li>在累加的最小值处进行切割，保留前面的。</li>
</ol>
<ul>
<li>概览</li>
</ul>
<p>FASTQ(QC, mapping-找reads在基因组的位置)- SAM（sequence alignment)- BAM（压缩的sam文件）</p>
<p>bam-质控（比对情况），计数，标准化，找差异表达</p>
<p>改进：翻译效率的问题，降解未结合核糖体的mRNA区域。 RNA结合蛋白鉴定：fastq-bam-chip-seq 待写。</p>
</section>
<section id="mapping" class="level4">
<h4 class="anchored" data-anchor-id="mapping">mapping</h4>
<ul>
<li><p>alignment（低通量，两条/多序列） pairwise multiple</p>
<p>pairwise: 全局，局部（needleman-wunch, smith-waterman）</p>
<p>blast: one vs many 序列切短，相似再扩展。</p></li>
<li><p>mapping（reads 回溯基因组） many vs one(bwt算法-bwa bowtie)</p>
<ol type="1">
<li>mapping回成熟的mRNA参考序列，不用处理可变剪切，不能发现新的转录本。</li>
<li>mapping到参考基因组，可以发现新的转录本，进行isoform层次的定量，但是不能使用之前的DNA mapping软件。</li>
</ol></li>
</ul>
<p>mapping的可变剪切问题：</p>
<ol type="1">
<li>exon-first approach(pseudogene-mRNA逆转录cDNA插入基因组，贴到假基因区)</li>
<li>seed-extend approch</li>
<li>potential limitations of exon-first approches。</li>
</ol>
<p>参考基因组(序列）下载：</p>
<p>UCSC genome broswer/download/human/sequence data by chromsome</p>
<p>合并染色体序列信息：<br>
chr2.fa.gz <code>cat chr1.fa.gz chr2.fa.gz &gt; ref_hg38.fa</code></p>
<p>&gt;chr1<br>
NNNNNNNNNN(端粒的占位符)<br>
ATCGGGGGGGG<br>
&gt;chr2<br>
NNNNNNNNNN<br>
ATCGGGGGGGG</p>
<p>Ensembl: ensembl species list/human/gene assemble/download DNA sequence<br>
NCBI: refseq/</p>
<p>参考基因注释文件（GTF,GFF-序列哪些位置是基因等)：<br>
ussc/tools/table browser</p>
<p>注释文件每列含义：待写</p>
<p>Note：注释文件和参考基因组match</p>
<p><strong>BWT(burrows wheeler transform)</strong></p>
<p>假定：ref- ACAACG<br>
ACAACG&amp; 循环</p>
<ol type="1">
<li>&amp;ACAACG<br>
G&amp;ACAAC<br>
CG&amp;ACAA<br>
ACG&amp;ACA<br>
AACG&amp;AC<br>
CAACG&amp;A</li>
<li>得到的字符矩阵根据第一列首字母排序（&amp;ACGT)，得到排序后的字符矩阵。</li>
<li>index就是排序的字符矩阵的最后一列</li>
</ol>
<p>排序后的字符矩阵有下面两个性质：</p>
<ul>
<li>每一行的第一个字符和最后一个字符一定在原始字符串的相连；且后一个字符在序列中在第一字符前。</li>
<li>第一列和最后一列字母（例如同一个A..)的相对次序不变；也就是第一列中第一个A和最后一列的第一个A实际上是序列中的同一个A。</li>
</ul>
<p>Index在mapping中的作用如下：</p>
<ol type="1">
<li>根据index还原排序后的字符矩阵矩阵的第一列</li>
<li>根据上面两条性质反推出reference。</li>
<li>根据排序矩阵第一列和最后一列比对，从序列后面，矩阵第一列开始搜索</li>
</ol>
<p><strong>后缀树算法- suffix tree</strong></p>
<ol type="1">
<li>同bwt得到未排序的字符矩阵。</li>
<li>按第一列首字符排序得到排序的字符矩阵。</li>
<li>构建suffix tree（存储树信息以及相对位置 信息，index索引非常大）</li>
<li>从树的根节点出发。</li>
</ol>
<p>实际上后缀树存储了序列某个字符后的所有可能性，例如某一字符为A，后缀树就穷举了序列中A后面所有字符的可能并保存下来，这种做法就是用空间换时间。</p>
<p>一些mapping软件：</p>
<p>tophat/tophat2 构建索引： <code>bowtie2-build --threds 6 ref_hg38.fa ref_hg38.fa</code></p>
<p>star-基于后缀树</p>
<ol type="1">
<li>reads切成小的seed，找到seed位置</li>
<li>符合的seed拼一起</li>
</ol>
<p>hisat2(tophat的升级版) - 全基因组Index，切割为55000份建立index。先定位在哪个小的index，然后在短的index搜索。（两层的bwt结构；其优点在于考虑了SNP信息（mapping时可以替换）</p>
<p>构建hisat2 index： SNP信息：dbSNP commom（ucsc genome/annotation/sql/common.txt.gz)<br>
可变剪切信息：GTF<br>
参考基因组信息：Genome FASTA</p>
<p>sam flag信息</p>
<p><code>samtool sort</code> PG bam（哪些操作）</p>
<p>bam 文件建立索引方便查看mapping信息。 <code>samtool index</code>，任意一段 <code>samtool view -h</code></p>
</section>
<section id="rna-seq-定量" class="level4">
<h4 class="anchored" data-anchor-id="rna-seq-定量">RNA seq 定量</h4>
<section id="通过参考基因组进行定量" class="level6">
<h6 class="anchored" data-anchor-id="通过参考基因组进行定量">通过参考基因组进行定量</h6>
<p><strong>样本内标准方法</strong></p>
<p>除以测序深度（reads数目），基因长度（bp)，进行单位化，分子乘以<span class="math inline">\(10^6\)</span>（每百万reads)， 乘以<span class="math inline">\(10^3\)</span>(基因长度1000bp)，1单位rpkm含义就是每百万reads中长度为1kb的基因的reads数目为1。</p>
<p>r-reads, f-fragments, p-per, k-kilobase, m-million。</p>
<p>RPM or CPM：仅对测序深度进行单位化。 <span class="math display">\[
\frac{Number\ of\ reads\ mapped\ to\ gene *10^6}{Total\ number\ of\ mapped\ reads}
\]</span> RPKM, FPKM：同时测序深度以及基因长度进行单位化。对于双端测序，fragment就是同一对reads。单端测序FPKM等于RPKM。 <span class="math display">\[
\frac{Number\ of\ reads\ mapped\ to\ gene* 10^6 *10^3}{Total\ number\ of\ mapped\ reads* gene\ length\ in\ bp}
\]</span></p>
<p>TPM:<br>
TPM假设每个样本的基因数值加和相同，其简单的为样本内基因Fpkm 的百分数*<span class="math inline">\(10^6\)</span>。</p>
<p><strong>样本间标准化</strong></p>
<p>直接计算比例： <span class="math display">\[
C_j = \frac{10^6}{D_j}
\]</span> 其中<span class="math inline">\(D_j\)</span>为样本<span class="math inline">\(j\)</span>测序深度，<span class="math inline">\(C_j\)</span>为样本<span class="math inline">\(j\)</span>校正系数。这种方法容易受到极端值的影响。</p>
<p>quantile： <span class="math display">\[
C_j = \frac{exp\left( \frac{1}{N}\sum_{l = 1}^{N}log(D_{l}Q_{l}^{(p)})\right)}{D_{j}Q_{j}^{(p)}}
\]</span></p>
<p>样本的分位数均值与某一个样本的分位数比值。其中<span class="math inline">\(D_j\)</span>为样本<span class="math inline">\(j\)</span>的测序深度，<span class="math inline">\(Q_j^{(p)}\)</span>为样本<span class="math inline">\(j\)</span>的<span class="math inline">\(p\)</span>分位数，<span class="math inline">\(C_j\)</span>为样本<span class="math inline">\(j\)</span>的校正系数。</p>
<p>RLE(relative log expression) - cufdiff；Deseq2默认方法：</p>
<p>假定行为基因，列为样本。</p>
<ol type="1">
<li>每行的几何平均数。</li>
<li>每行除以该行的几何平均数，得到新矩阵。</li>
<li>新矩阵每列的中位数就是该列样本的校正因子。</li>
</ol>
<p><span class="math display">\[
C_j = median_g\left( \frac{K_{gj}}{(\prod_{l=1}^{N}K_{gl})^{\frac{1}{N}}} \right)
\]</span></p>
<p>TMM(edge R) Trimmed mean of M-values：</p>
<p>假设前提：total reads受高表达基因影响。大多数基因的表达量不变。</p>
<p>RNA-seq的定量在MA plot反应为，1）大多数点应该贴近于M = 0该直线附近。2）高表达基因的应该贴近于M = 0该直线附近———横轴是按基因的几何平均由低（左）到高（右边）排布的。</p>
<p>MA plot： X-A：两组样本的几何平均数 <span class="math display">\[
A = \frac{1}{2}log_2{(RG)}
\]</span> M-Y：两组样本的fold change。 <span class="math display">\[
M = log_2{(R/G)}
\]</span></p>
<p>变化基因以及高表达的基因的阈值可选。</p>
</section>
</section>
<section id="rsem转录水平定量" class="level4">
<h4 class="anchored" data-anchor-id="rsem转录水平定量">RSEM转录水平定量</h4>
<p>reads直接mapping到mRNA上，解决可变剪切的问题。reads来自于哪个isoform?从极大似然估计到EM算法，可以得出每个isoform的count数目。</p>
</section>
</section>
<section id="rnaseq下游分析篇" class="level2">
<h2 class="anchored" data-anchor-id="rnaseq下游分析篇">RNASeq下游分析篇</h2>
<section id="差异分析" class="level4">
<h4 class="anchored" data-anchor-id="差异分析">差异分析</h4>
<p>基因的定量是一个抽样的结果RNA-cDNA-RNA</p>
<p>RNA-Seq的前提：</p>
<ul>
<li><p>绝大多数基因的表达量不变</p></li>
<li><p>高表达基因的表达量不变</p></li>
<li><p>如果需要绝对定量，使用提前绝对定量的内参（spike-ins)，ERCC countrol。</p></li>
</ul>
<p>对于cell line进行差异分析，需要2-3个repeat，对于生物体进行差异分析需要3-个repeat，对于群体而言，需要成百上千个repeate。</p>
<ul>
<li>p-value校正：</li>
</ul>
<p>p排序<span class="math inline">\(\rightarrow\)</span> 新p = p* 检验次数m <span class="math inline">\(\rightarrow\)</span>保持原有顺序（第一次排序的结果）调整顺序</p>
<p>BH校正 ：p排序<span class="math inline">\(\rightarrow\)</span>新p = p * 检验次数 / 序号<span class="math inline">\(\rightarrow\)</span>保持原有顺序调整p值。</p>
<ul>
<li>RNA-Seq定量的分布模型</li>
</ul>
<p><strong>RNA-Seq定量的二项分布</strong>：考虑一个基因，其它所有基因为一部分。那么基因A的出现次数二项分布。</p>
<p><strong>RNA-Seq定量的泊松分布</strong>：因为一个基因出现的概率很低，p很低，二项分布接近于泊松分布。</p>
<p><strong>RNA-Seq定量的负二项分布</strong>：</p>
<p>实际上RNA-seq的数据并不服从泊松分布，泊松分布期望与方差相等，然而RNA-Seq图像是over-dispersion的。随均值增加，方差也变大。这种现象称为short noise。</p>
<p>泊松分布： <span class="math display">\[
E(K_g) = Var(K_g) = \lambda
\]</span></p>
<p>RNA-seq short noise现像通过对泊松分布的修正描述。</p>
<p><span class="math display">\[
E(K_g) = \lambda, \ Var(K_g) = \lambda + \phi\lambda^2
\]</span></p>
<p>负二项分布具有此特性。所有RNA-seq基因的分布修正为负二项分布。</p>
<p>差异检验的零假设： <span class="math display">\[
\lambda_{g}A = \lambda_{g}B
\]</span></p>
<p>问题变成了估计<span class="math inline">\(\lambda_g\)</span>，<span class="math inline">\(\phi_g\)</span>，这个过程叫做estimate dispersion，不同软件的估算方法不同。</p>
<p>以前的统计检验思路，列联表卡方检验或fisher检验。基因同分布检验，fisher对大数字敏感，对小数字不敏感。</p>
<ul>
<li><p>RNA-seq的绝对定量</p></li>
<li><p>ERCC 建库加入定量的已知ERCC spike-in mRNA</p></li>
<li><p>house kepping(3000-4000）基因定量 输入文件是 未经过校正的count数据，认为不变的基因list(spike in) 输出就是校正过的数据。</p></li>
</ul>
</section>
</section>
<section id="基因注释与富集分析" class="level2">
<h2 class="anchored" data-anchor-id="基因注释与富集分析">基因注释与富集分析</h2>
<p>检验：GO kegg注释就是列联表的卡方检验。<br>
输入：一个感兴趣基因集，基因注释信息。<br>
输出：基因基是否在某一类注释信息中。<br>
需要认为设定感兴趣基因集的阈值。</p>
<p>GESA：解决人为设定基因集的问题。 输入：全部的基因变化信息，一个感兴趣的通路的基因集合（GESA msi)。<br>
输出：是否与整个感兴趣的通路相关。</p>
<p>图片认知（待写）</p>
<p>非模式生物的富集分析：有参，无参。<br>
annotationhub。</p>
<section id="多样本数据分析" class="level4">
<h4 class="anchored" data-anchor-id="多样本数据分析">多样本数据分析</h4>
<p>TCGA是肿瘤数据库，可以通过firebrowser下载其中的数据。GTex是正常人的数据。</p>
<p>关于pearson相关系数以及spearman相关系数：spearman相关系数仅仅是对数据的排序序号进行pearson相关分析的结果。</p>
<p>WGCNA：只是简单对基因进行聚类，通过最终聚类效果的评价指标选取距离计算的指数。</p>
<p>多样本差异表达基因：</p>
<p>将负二项模型通过对数连接函数写作广义线性模型，并对其参数进行似然比检验：</p>
<p><span class="math display">\[
log\mu_{gi} = x_{i}^{t}\beta_g + logN_i
\]</span></p>
<p>其中<span class="math inline">\(log\mu_{gi}\)</span>是基因<span class="math inline">\(g\)</span>在样本<span class="math inline">\(i\)</span>的观测值，<span class="math inline">\(x_{i}^{t}\)</span>设计矩阵，<span class="math inline">\(N_i\)</span>是样本的测序深度。差异检验就是对<span class="math inline">\(\beta_g\)</span>是否全为0的检验。</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>